ooni/agent/scheduler.py

```diff
@@ -0,0 +1,423 @@
+import os
+import errno
+
+from hashlib import md5
+from datetime import datetime
+
+from twisted.application import service
+from twisted.internet import defer
+from twisted.internet.task import LoopingCall
+from twisted.python.filepath import FilePath
+
+from ooni.scripts import oonireport
+from ooni import resources
+from ooni.utils import log, SHORT_DATE
+from ooni.utils.files import human_size_to_bytes, directory_usage
+from ooni.deck.store import input_store, deck_store, DEFAULT_DECKS
+from ooni.settings import config
+from ooni.contrib import croniter
+from ooni.contrib.dateutil.tz import tz
+
+from ooni.geoip import probe_ip
+from ooni.measurements import list_measurements
+
+class FileSystemlockAndMutex(object):
+    """
+    This is a lock that is both a mutex lock and also on filesystem.
+    When you acquire it, it will first acquire the mutex lock and then
+    acquire the filesystem lock. The release order is inverted.
+
+    This is to avoid concurrent usage within the same process. When using it
+    concurrently the mutex lock will block before the filesystem lock is
+    acquired.
+
+    It's a way to support concurrent usage of the DeferredFilesystemLock
+    without races.
+    """
+    def __init__(self, file_path):
+        self._fs_lock = defer.DeferredFilesystemLock(file_path)
+        self._mutex = defer.DeferredLock()
+
+    @defer.inlineCallbacks
+    def acquire(self):
+        yield self._mutex.acquire()
+        yield self._fs_lock.deferUntilLocked()
+
+    def release(self):
+        self._fs_lock.unlock()
+        self._mutex.release()
+
+    @property
+    def locked(self):
+        return self._mutex.locked or self._fs_lock.locked
+
+# We use this date to indicate that the scheduled task has never run.
+# Easter egg, try to see what is special about this date :)?
+CANARY_DATE = datetime(1957, 8, 4, tzinfo=tz.tzutc())
```

Comment:
Previously we would have had epoch 0 in here, but I thought it would be better to have a date that is actually specific to ooniprobe to distinguish it from a time that may be set when a system boots up with 0 seconds from epoch and hence leading to a timestamp of 0 (that would mean we would trigger the first_run) twice.

Since we manage the time always ourselves and serialize it using methods that support times before epoch we should be all good.